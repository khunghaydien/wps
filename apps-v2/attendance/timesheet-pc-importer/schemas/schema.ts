import * as yup from 'yup';

import * as messages from './messages';

yup.setLocale(messages);

export default yup;
