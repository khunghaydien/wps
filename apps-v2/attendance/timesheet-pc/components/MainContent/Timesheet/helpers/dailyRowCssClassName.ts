import {
  isWithDeduction,
  isWithPayment,
  LeaveType,
} from '@attendance/domain/models/LeaveType';
import AttRecordModel from '@attendance/timesheet-pc/models/AttRecord';

export default (classParam: {
  ROOT: string;
  record: AttRecordModel;
  effectualAllDayLeaveType: LeaveType | null;
  isManHoursGraphOpened: boolean;
  isApprovedAbsence: boolean;
  isAllowWorkDuringLeaveOfAbsence: boolean;
}) => {
  const {
    ROOT,
    record,
    effectualAllDayLeaveType,
    isManHoursGraphOpened,
    isApprovedAbsence,
    isAllowWorkDuringLeaveOfAbsence,
  } = classParam;

  return {
    [`${ROOT}--man-hours-graph-opened`]: isManHoursGraphOpened,

    [`${ROOT}--day-type-legal-holiday`]:
      record.dayType === AttRecordModel.DAY_TYPE.LEGAL_HOLIDAY ||
      record.isHolLegalHoliday === true,
    [`${ROOT}--day-type-workday`]:
      record.dayType === AttRecordModel.DAY_TYPE.WORKDAY,
    [`${ROOT}--day-type-holiday`]:
      record.dayType === AttRecordModel.DAY_TYPE.HOLIDAY ||
      record.dayType === AttRecordModel.DAY_TYPE.PREFERRED_LEGAL_HOLIDAY,

    [`${ROOT}--leave-with-payment`]:
      effectualAllDayLeaveType !== null &&
      isWithPayment(effectualAllDayLeaveType),
    [`${ROOT}--leave-with-deduction`]:
      effectualAllDayLeaveType !== null &&
      isWithDeduction(effectualAllDayLeaveType),

    [`${ROOT}--leave-of-absence`]: record.isLeaveOfAbsence,

    [`${ROOT}--absence`]: isApprovedAbsence,

    [`${ROOT}--work-during-leave-of-absence`]: isAllowWorkDuringLeaveOfAbsence,
    [`${ROOT}--work-during-leave-of-absence-with-payment`]:
      effectualAllDayLeaveType !== null &&
      isWithPayment(effectualAllDayLeaveType),
    [`${ROOT}--work-during-leave-of-absence-with-deduction`]:
      effectualAllDayLeaveType !== null &&
      isWithDeduction(effectualAllDayLeaveType),
  };
};
