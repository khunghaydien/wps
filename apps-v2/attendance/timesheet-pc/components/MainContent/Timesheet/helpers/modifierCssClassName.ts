import classNames from 'classnames';

import { TimesheetViewType } from '../TimesheetViewType';

export default (
  baseName: string,
  type: TimesheetViewType,
  {
    useFixDailyRequest,
  }: {
    useFixDailyRequest: boolean;
  } = {
    useFixDailyRequest: false,
  }
) =>
  classNames(`${baseName}--${type}`, {
    [`${baseName}--use-fix-daily-request`]: useFixDailyRequest,
  });
