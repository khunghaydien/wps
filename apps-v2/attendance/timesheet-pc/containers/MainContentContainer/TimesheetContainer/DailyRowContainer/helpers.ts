import isEqual from 'lodash/isEqual';

import { DailyRecordDisplayFieldLayoutItemValueForUI } from '@attendance/timesheet-pc/modules/ui/dailyRecordDisplayFieldLayout';

export const fieldChanged = ({
  orignValues,
  currentValues,
}: {
  orignValues: {
    [itemId: string]: DailyRecordDisplayFieldLayoutItemValueForUI;
  };
  currentValues: {
    [itemId: string]: DailyRecordDisplayFieldLayoutItemValueForUI;
  };
}): boolean => {
  return !isEqual(orignValues, currentValues);
};
