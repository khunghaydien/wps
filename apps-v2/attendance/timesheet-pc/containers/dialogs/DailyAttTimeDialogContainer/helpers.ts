import AttRecord from '../../../models/AttRecord';
import { DailyObjectivelyEventLog } from '@attendance/domain/models/DailyObjectivelyEventLog';
import { MAX_STANDARD_REST_TIME_COUNT } from '@attendance/domain/models/RestTime';

import compare from '@apps/attendance/libraries/utils/compare';

import * as DailyAttTimeViewModel from '@attendance/timesheet-pc/viewModels/EditingDailyAttendanceTimeViewModel';
import {
  DailyRecordFactory as ToDailyRecordFactory,
  Factory as DailyAttTimeViewModelFactory,
} from '@attendance/timesheet-pc/viewModels/EditingDailyAttendanceTimeViewModel';

export const convertToSaving = ToDailyRecordFactory.create;

export const convertToEditing = DailyAttTimeViewModelFactory.createByAttRecord;

export const isChange = ({
  dailyAttTime,
  attRecord,
  dailyObjectivelyEventLog,
}: {
  dailyAttTime: DailyAttTimeViewModel.EditingDailyAttendanceTimeViewModel;
  attRecord: AttRecord;
  dailyObjectivelyEventLog: DailyObjectivelyEventLog | null;
}): boolean => {
  const originalRecord = convertToSaving(
    convertToEditing(
      attRecord,
      MAX_STANDARD_REST_TIME_COUNT,
      dailyObjectivelyEventLog
    )
  );
  const record = convertToSaving(dailyAttTime);
  return !compare(record, originalRecord);
};
