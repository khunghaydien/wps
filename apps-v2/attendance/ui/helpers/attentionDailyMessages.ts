import msg from '@apps/commons/languages';
import TextUtil from '@apps/commons/utils/TextUtil';
import TimeUtil from '@apps/commons/utils/TimeUtil';

import {
  AttDailyAttention,
  CODE,
} from '@attendance/domain/models/AttDailyAttention';

export const tip = (attentions: AttDailyAttention[]): string | null => {
  const messages = alert(attentions);
  return !messages
    ? null
    : messages.length === 1
    ? messages[0]
    : msg().Att_Msg_MultipulAttentionMessage;
};

export const alert = (attentions: AttDailyAttention[]): string[] | null => {
  const messages = attentions
    .map((attention) => {
      switch (attention.code) {
        case CODE.INEFFECTIVE_WORKING_TIME:
          return TextUtil.template(
            msg().Att_Msg_NotIncludeWorkingTime,
            TimeUtil.toHHmm(attention.value.fromTime),
            TimeUtil.toHHmm(attention.value.toTime)
          );
        case CODE.INSUFFICIENT_REST_TIME:
          return TextUtil.template(
            msg().Att_Msg_InsufficientRestTime,
            attention.value
          );
        case CODE.OVER_ALLOWING_DEVIATION_TIME:
          return msg().Att_Msg_OverAllowingDeviationTime;
        case CODE.OUT_INSUFFICIENT_MINIMUM_WORK_HOURS:
          return TextUtil.template(
            msg().Att_Msg_OutInsufficientMinimumWorkHours,
            TimeUtil.toHHmm(attention.value)
          );
        default:
          return null;
      }
    })
    .filter((m) => m);
  return messages.length ? messages : null;
};

export const remarks = (attentions: AttDailyAttention[]): string[] | null => {
  const messages = attentions
    .map((attention) => {
      switch (attention.code) {
        case CODE.INEFFECTIVE_WORKING_TIME:
          return TextUtil.template(
            msg().Att_Msg_SummaryCommentIneffectiveWorkingTime,
            TimeUtil.toHHmm(attention.value.fromTime),
            TimeUtil.toHHmm(attention.value.toTime)
          );
        case CODE.INSUFFICIENT_REST_TIME:
          return TextUtil.template(
            msg().Att_Msg_SummaryCommentInsufficientRestTime,
            attention.value
          );
        // NOTE: 必要ならば追加してください。
        // case CODE.OverAllowingDeviationTime:
        default:
          return null;
      }
    })
    .filter((m) => m);
  return messages.length ? messages : null;
};
