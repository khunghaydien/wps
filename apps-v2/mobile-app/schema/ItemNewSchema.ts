import * as Yup from 'yup';

import msg from '../../commons/languages';

import getExtendedItemsValidators from './expenses/ExtendedItemSchema';

export default Yup.lazy(() => {
  return Yup.object().shape({
    items: Yup.array().of(
      Yup.object().shape({
        amount: Yup.number()
          .typeError(msg().Common_Err_TypeNumber)
          .required(msg().Common_Err_Required)
          .max(999999999999, msg().Common_Err_Max)
          .nullable(),
        recordDate: Yup.string().required(msg().Common_Err_Required).nullable(),
        expTypeId: Yup.string()
          .nullable()
          .test(
            'is-required',
            msg().Common_Err_Required,
            function validateExpTypeId(value) {
              if (!this.path.includes('items[0]')) {
                return !!value;
              }
            }
          ),
        expTypeName: Yup.string()
          .required(msg().Common_Err_Required)
          .nullable(),
        useForeignCurrency: Yup.boolean(),
        taxTypeBaseId: Yup.string()
          .nullable()
          .when('useForeignCurrency', {
            is: false,
            then: Yup.string().required(msg().Common_Err_Required),
          }),
        remarks: Yup.string().max(1024, msg().Common_Err_Max).nullable(),
        ...getExtendedItemsValidators(true),
      })
    ),
  });
});
