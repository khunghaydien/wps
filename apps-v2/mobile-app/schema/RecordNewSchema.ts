import get from 'lodash/get';
import * as Yup from 'yup';

import { amountMatchSchema } from '@commons/schema/Expense';

import msg from '../../commons/languages';

import {
  isMileageRecord,
  isUseWithholdingTax,
  RECEIPT_TYPE,
  RECORD_TYPE,
  WITHHOLDING_TAX_TYPE,
} from '../../domain/models/exp/Record';

import getExtendedItemsValidators from './expenses/ExtendedItemSchema';
import {
  paymentDueDateRecordSchema,
  vendorIdRecordSchema,
} from './request/RecordNewSchema';

function validateAttachmentBody(value) {
  const hasRemarks =
    this.parent.items[0].remarks && this.parent.items[0].remarks.trim();
  if (hasRemarks || (value && value.length > 0)) {
    return true;
  }
  return false;
}

export const amountNegativeSchema = (msgKey) => {
  return Yup.number().when('allowNegativeAmount', {
    is: false,
    then: Yup.number().min(0, msgKey),
  });
};

export const amountPayableSchema = (values) =>
  Yup.number()
    .nullable()
    .when('.', {
      is: () => {
        const withholdingTaxUsage = get(values, 'withholdingTaxUsage');
        return isUseWithholdingTax(withholdingTaxUsage);
      },
      then: Yup.number().min(0, msg().Exp_Err_AmountPayableNegative),
    });

export default Yup.lazy((values) => {
  return Yup.object().shape({
    amount: amountMatchSchema(),
    fileAttachment: Yup.string().nullable(),
    receiptList: Yup.array()
      .when('fileAttachment', {
        is: RECEIPT_TYPE.Required,
        then: Yup.array().min(1, msg().Common_Err_Required),
      })
      .when('fileAttachment', {
        is: RECEIPT_TYPE.Optional,
        then: Yup.array()
          .nullable()
          .test(
            'GeneralReceiptOptional',
            msg().Exp_Err_NoReceiptAndReceipt,
            validateAttachmentBody
          ),
      }),
    items: Yup.array().of(
      Yup.object().shape({
        recordDate: Yup.string().required(msg().Common_Err_Required).nullable(),
        amount: amountNegativeSchema('Exp_Err_AmountNegative')
          .typeError(msg().Common_Err_TypeNumber)
          .required(msg().Common_Err_Required)
          .max(999999999999, msg().Common_Err_Max)
          .nullable(),
        localAmount: Yup.number().when('useForeignCurrency', {
          is: true,
          then: amountNegativeSchema('Exp_Err_LocalAmountNegative'),
        }),
        expTypeName: Yup.string()
          .required(msg().Common_Err_Required)
          .nullable(),
        useForeignCurrency: Yup.boolean(),
        taxTypeBaseId: Yup.string()
          .nullable()
          .when(['useForeignCurrency', 'taxItems'], {
            is: (useForeignCurrency, taxItems) => {
              const requireTaxValidation = !useForeignCurrency && !taxItems;
              return requireTaxValidation;
            },
            then: Yup.string().required(msg().Common_Err_Required),
          }),
        merchant: Yup.string()
          .nullable()
          .test(
            'is-required',
            msg().Common_Err_Required,
            function validateNoMerchant(value) {
              if (!this.path.includes('items[0]')) {
                return true;
              }
              const merchantUsage = get(values, 'merchantUsage');
              const isRequired = merchantUsage === RECEIPT_TYPE.Required;
              if (isRequired && !value) {
                return false;
              }
              return true;
            }
          ),
        mileageDistance: Yup.number()
          .nullable()
          .test('is-required', 'Com_Lbl_Required', function (distance: number) {
            const isMileage = isMileageRecord(values.recordType);
            return isMileage ? Math.abs(distance) > 0 : true;
          }),
        fixedAllowanceOptionId: Yup.string()
          .when({
            is: values && values.recordType === RECORD_TYPE.FixedAllowanceMulti,
            then: Yup.string().nullable().required(msg().Common_Err_Required),
          })
          .nullable(),
        remarks: Yup.string().max(1024, msg().Common_Err_Max).nullable(),
        amountPayable: amountPayableSchema(values),
        withholdingTaxAmount: Yup.number()
          .nullable()
          .test(
            'is-required',
            msg().Common_Err_Required,
            function (withholdingTaxAmount: number) {
              if (!this.path.includes('items[0]')) return true;
              const withholdingTaxUsage = get(values, 'withholdingTaxUsage');
              const isRequired =
                withholdingTaxUsage === WITHHOLDING_TAX_TYPE.Required;
              return isRequired ? Math.abs(withholdingTaxAmount) > 0 : true;
            }
          ),
        vendorId: vendorIdRecordSchema(values),
        paymentDueDate: paymentDueDateRecordSchema(values),
        ...getExtendedItemsValidators(true),
      })
    ),
    mileageRouteInfo: Yup.object()
      .when('recordType', {
        is: RECORD_TYPE.Mileage,
        then: Yup.object().shape({
          destinations: Yup.array().of(
            Yup.object().shape({
              name: Yup.string().required('Exp_Err_Destination'),
            })
          ),
        }),
      })
      .nullable(),
    transitIcRecordId: Yup.string()
      .when('recordType', {
        is: RECORD_TYPE.TransportICCardJP,
        then: Yup.string()
          .nullable()
          .required(msg().Exp_Err_NoIcTransactionLinked),
      })
      .nullable(),
    creditCardTransactionId: Yup.string()
      .nullable()
      .test(
        'creditCardTransactionId',
        msg().Exp_Err_NoCcTransactionLinked,
        (creditCardTransactionId: string) => {
          if (values.isCCPaymentMethod) return !!creditCardTransactionId;
          return true;
        }
      )
      .nullable(),
  });
});
