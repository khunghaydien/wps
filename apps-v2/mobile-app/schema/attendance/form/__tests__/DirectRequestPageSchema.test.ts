import schema from '../DirectRequestPageSchema';
import { create } from './helpers/errors';

const validate = create(schema);

it('return error when data is empty.', async () => {
  await expect(validate({})).rejects.toMatchSnapshot();
});

it('return error when data is null.', async () => {
  await expect(
    validate({
      startDate: null,
      endDate: null,
      startTime: null,
      endTime: null,
      remarks: null,
    })
  ).rejects.toMatchSnapshot();
});

it("return error when data is ''.", async () => {
  await expect(
    validate({
      startDate: '',
      endDate: '',
      startTime: '',
      endTime: '',
      remarks: '',
    })
  ).rejects.toMatchSnapshot();
});

it('return error when data is not type.', async () => {
  await expect(
    validate({
      startDate: 100,
      endDate: 100,
      startTime: 'abc',
      endTime: 'abc',
      remarks: 100,
    })
  ).rejects.toMatchSnapshot();
});

it('return true.', async () => {
  await expect(
    validate({
      startDate: 'abc',
      endDate: 'abc',
      startTime: 0,
      endTime: 0,
      remarks: '',
    })
  ).resolves.toBeTruthy();
});
