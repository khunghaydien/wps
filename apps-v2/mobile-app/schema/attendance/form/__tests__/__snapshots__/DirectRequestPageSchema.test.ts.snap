// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return error when data is ''. 1`] = `
Object {
  "endTime": Array [
    "endTime must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"\\"\`).",
  ],
  "startDate": Array [
    " Required",
  ],
  "startTime": Array [
    "startTime must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"\\"\`).",
  ],
}
`;

exports[`return error when data is empty. 1`] = `
Object {
  "endTime": Array [
    " Required",
  ],
  "startDate": Array [
    " Required",
  ],
  "startTime": Array [
    " Required",
  ],
}
`;

exports[`return error when data is not type. 1`] = `
Object {
  "endTime": Array [
    "endTime must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"abc\\"\`).",
  ],
  "startTime": Array [
    "startTime must be a \`number\` type, but the final value was: \`NaN\` (cast from the value \`\\"abc\\"\`).",
  ],
}
`;

exports[`return error when data is null. 1`] = `
Object {
  "endTime": Array [
    " Required",
  ],
  "startDate": Array [
    " Required",
  ],
  "startTime": Array [
    " Required",
  ],
}
`;
